src/jit/gentree.cpp

```diff
@@ -6586,115 +6799,165 @@ GenTreePtr Compiler::gtNewAssignNode(GenTreePtr dst, GenTreePtr src)
-
-        if (size >= TARGET_POINTER_SIZE)
+        GenTree* val = addr->gtGetOp1();
+        if (val->OperGet() == GT_LCL_VAR)
         {
-            slots  = (unsigned)(roundUp(size, TARGET_POINTER_SIZE) / TARGET_POINTER_SIZE);
-            gcPtrs = new (this, CMK_ASTNode) BYTE[slots];
-
-            type = impNormStructType(structHnd, gcPtrs, &gcPtrCount);
-            if (varTypeIsEnregisterableStruct(type))
+            unsigned   lclNum = addr->gtGetOp1()->AsLclVarCommon()->gtLclNum;
+            LclVarDsc* varDsc = &(lvaTable[lclNum]);
+            if (varTypeIsStruct(varDsc) && (varDsc->lvVerTypeInfo.GetClassHandle() == structHnd) &&
+                !lvaIsImplicitByRefLocal(lclNum))
             {
-                if (dst->OperGet() == GT_ADDR)
-                {
-                    GenTree* actualDst = dst->gtGetOp1();
-                    assert((actualDst->TypeGet() == type) || !varTypeIsEnregisterableStruct(actualDst));
-                    actualDst->gtType = type;
-                }
-                if (src->OperGet() == GT_ADDR)
-                {
-                    GenTree* actualSrc = src->gtGetOp1();
-                    assert((actualSrc->TypeGet() == type) || !varTypeIsEnregisterableStruct(actualSrc));
-                    actualSrc->gtType = type;
-                }
+                return addr->gtGetOp1();
             }
+        }
+    }
+    return gtNewObjNode(structHnd, addr);
+}
 
-            useCopyObj = gcPtrCount > 0;
+//------------------------------------------------------------------------
+// gtNewBlockVal: Return a node that represents a possibly untyped block value
+//
+// Arguments:
+//    addr      - The address of the block
+//    size      - The size of the block
+//
+// Return Value:
+//    A block, object or local node that represents the block value pointed to by 'addr'.
+
+GenTree* Compiler::gtNewBlockVal(GenTreePtr addr, unsigned size)
+{
+    // By default we treat this as an opaque struct type with known size.
+    var_types blkType = TYP_STRUCT;
+#if FEATURE_SIMD
+    if ((addr->gtOper == GT_ADDR) && (addr->gtGetOp1()->OperGet() == GT_LCL_VAR))
+    {
+        GenTree* val = addr->gtGetOp1();
+        if (varTypeIsSIMD(val) && (genTypeSize(val->TypeGet()) == size))
+        {
+            blkType = val->TypeGet();
+            return addr->gtGetOp1();
         }
     }
+#endif // FEATURE_SIMD
+    return new (this, GT_BLK) GenTreeBlk(GT_BLK, blkType, addr, size);
+}
 
-    // If the class being copied contains any GC pointer we store a class handle
-    // in the icon, otherwise we store the size in bytes to copy
-    //
-    genTreeOps op;
-    if (useCopyObj)
+// Creates a new assignment node for a CpObj.
+// Parameters (exactly the same as MSIL CpObj):
+//
+//  dst        - The target to copy the struct to
+//  src        - The source to copy the struct from
+//  structHnd  - A class token that represents the type of object being copied. May be null
```

Comment:
Yes, good point. I had attempted to make this take actual values, but it was difficult to preserve existing behavior. In future this may be changed to take the struct values, but for now I'll do as you suggest and change the names.

---
In reply to: [77387710](https://github.com/dotnet/coreclr/pull/7027#discussion_r77387710) [](ancestors = 77387710)